### PR TITLE
feat(ci): add portability-lint gate for ecosystem/forge-agnostic skills (#531)

### LINKED ISSUE #620: portability lint: branch-class gate engine — origin/(main|master) token class + engine skeleton (child of #531, delivered by PR #609)
Part of #531 — child issue for the scope PR #609 actually ships (the #603 child-per-stage pattern; filed by the control tower at merge digest so the umbrella does not auto-close). STOPGAP NOTE: plain child pending the #552 sub-issues convention.

Shipped scope (PR #609): fail-closed portability-lint CI lane — self-test-first orchestrator (`scripts/check-skill-portability.sh` + `.test.sh`), external token file, `portability-ok:`/`portability-scope:` escape hatches, changed-file scoping, `ci-status.needs` wiring; ACTIVE token class = branch-resolution hardcodes `origin/(main|master)` with guard-vocabulary recognition (`origin/HEAD`, `symbolic-ref`, `merge-base`, `baseRefName`).

Not in this child (still owed by the re-scoped #612): the cross-plugin bare-ref class, presence-guard vocabulary (`if installed`/`dependencies`), #405/#412 pass fixtures. Known deferral: inline-`<!--` annotation carry (#611).

### PR DESCRIPTION
**Pull request by kyle-sexton**
## Summary

Skills declared ecosystem/forge/tracker-agnostic ship bare hardcoded
stack/forge/branch/tracker defaults because agnosticism was asserted in prose and
never enforced mechanically — the dominant review-churn class, re-caught by the
external reviewer PR after PR (branch/remote hardcodes, seam-conditional eval,
.NET-flavored examples). This adds a fail-closed CI lane that mechanically detects
that coupling instead of paying for it at review time again and again.

Detection only — this PR builds the gate. Fixing the existing violations stays
with the member issues below.

## Fix

A new `portability-lint` lane (`.github/workflows/ci.yml`) wired into the required
`ci-status` aggregate, plus a repo-local detector:

- **`scripts/check-skill-portability.sh`** — scans skill files for coupling tokens.
  On a PR it scans only the skill files the change **touches** (mirroring the
  `skill-quality-gate` changed-diff pattern), so enabling a token class prevents
  **new** coupling without red-lining pre-existing violations — main's push event
  scans nothing (self-test is the push path), and existing hits wait for their
  owning follow-up fix or the file's next edit.
- **`scripts/skill-portability-tokens.txt`** — the token list as external,
  extensible **data** (not logic buried in bash), so a reviewer re-catch is a
  one-line data edit. Seeded with **one active class** — the branch/default-branch
  hardcode (`origin/main` / `origin/master`) — per the ratified one-token-class-at-
  a-time rollout; further classes (`dotnet`/`Clean Arch`, `raw.githubusercontent`,
  bare `gh` tracker calls) are staged as commented entries with enable-triggers.
- **Never-skip shape**: the job is unconditional; only the PR-diff step is
  event-gated, and a self-test step runs first so a broken detector can never mask
  a real violation behind a green gate.

**Design decisions** (resolved per the ratified plan, not invented):

- *How a skill declares agnosticism scope* — **no new frontmatter field**. A skill
  is agnostic by default (the Design boundary already binds every plugin), so the
  gated set is the files a change touches. A hit is excused three ways, all
  reviewer-visible comments (reusing the silent-skip gate's annotated-exemption
  shape): an auto-recognized detection-first / presence-gated use; a per-site
  `portability-ok: <reason>`; or a whole-file `portability-scope: <reason>`
  declaring an inherent narrower boundary (the forge-locked-under-a-neutral-name
  case). This distinguishes **guarded** refs (fine) from **bare** ones, as the
  guarded-forward-ref disposition requires.
- `docs/PLUGIN-PHILOSOPHY.md` gains one doctrine sentence extending the existing
  declared-narrower-boundary allowance from OS platform to the
  forge/ecosystem/tracker axis.

No `plugins/<name>/` directory is touched, so no version bump / CHANGELOG entry is
needed (matching the CI-gate precedent).

## Verification

- **Self-test suite** (`scripts/check-skill-portability.test.sh`, runs in CI):
  `PASS=12 FAIL=0` — covers bare-token FAIL with file:line, detection-ladder pass,
  same-line and comment-block-above `portability-ok`, annotation non-leak past
  intervening code, whole-file `portability-scope`, staged tokens staying inactive
  under the shipped list, fail-closed exit 2 on missing token list / invalid base
  ref, the vendor/evals/`*.test.sh` exclusion set, and empty scope.

- **Catches a real violation, passes a real legit use** (the bare-vs-guarded
  discrimination that drives this class), on live corpus files:

  ```
  $ scripts/check-skill-portability.sh --paths plugins/work-items/skills/track/actions/start.md
  COUPLING: plugins/work-items/skills/track/actions/start.md:61: origin/(main|master) -> ... git checkout -b <type>/<N>-<slug> origin/main ...
  COUPLING: plugins/work-items/skills/track/actions/start.md:65: origin/(main|master) -> ... git checkout -b <type>/<N>-<slug> origin/main ...
  (exit 1)

  $ scripts/check-skill-portability.sh --paths plugins/review/skills/fanout/SKILL.md
  No unexcused coupling tokens in 1 skill file(s).   # detection-first origin/HEAD ladder → guarded
  (exit 0)
  ```

- **Calibrated against the full corpus** (`--all`): the active branch class flags
  exactly 3 genuine bare hardcodes in 2 files (owned by member issues; changed-file
  scoping keeps them off main) and correctly passes the detection-ladder uses in
  `review/*` and excludes evals/test fixtures.

- **Changed-file CI path proven** against a real base ref (16 changed skills gated,
  all clean). Local checks green: `shellcheck` (repo `.shellcheckrc`), `shfmt`,
  `actionlint`, comment-residue detector (T1=T2=T3=0), `typos`.

Closes #620

Part of #531 (umbrella stays open — stages remain; child #620 carries this PR's shipped scope per the #603 pattern).

## Related

- Member coupling instances (this is prevention; it does not block them): #404
  #405 #406 #408 #410 #412 #415 #416 #418 #421 #422 #423 #428 #429 #432 #438 #439
  #441 #442
- #445 — sibling CI-gate backlog (mechanical conformance only; explicitly not
  coupling)
- #412 — guarded-forward-ref disposition the gate honors (guarded ≠ bare)
- #441 — declared narrower-scope exemption the `portability-scope` mechanism serves
- #467 — the branch/remote review-churn the seed active class targets
- #453 — seam-conditional eval re-catch of this class
- #491 #492 — .NET-flavored-example hand-fixes a future staged class would obviate
- #611 — post-green review finding (annotation-carry pending_annot doesn't distinguish inline vs block HTML comments), deferred out of this PR's scope

🤖 Generated with [Claude Code](https://claude.com/claude-code)

https://claude.ai/code/session_01KyLj6oaFVTE6xFuoYCC2KC

### CODE REVIEW

**Comment by kyle-sexton** [comment]
## Review classification (work-2 lane, post-green pass)

| # | Finding | Classification | Disposition |
|---|---|---|---|
| 1 | `\b` in staged (inactive) token patterns is non-POSIX ERE (gawk extension); silent no-op on mawk/nawk if enabled without testing | VALID | Fixing inline — warning comment added above the staged patterns |
| 2 | Generic `is_guarded()` markers (`if using`, `when installed`, etc.) will over-match once `dotnet`/`gh`-tracker staged classes activate, masking real hardcodes | VALID | Fixing inline — reminder comment added to the staged-class blocks |
| 3 | `is_comment()` matches inline `<!-- -->` on a content line, so `pending_annot` can carry forward from a non-block-comment line | VALID, deferred | Filed #611 — low-likelihood edge case, not exercised by current active token class or self-test; fixing here would add script-logic + test scope beyond this PR's ship boundary |

Findings 1–2: fix in progress on this branch (comment-only additions, no behavior change). Finding 3: tracked in #611, cited in `## Related`.

File: scripts/check-skill-portability.sh [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+  ' "$TOKENS" "$file"
+}
+
+violations=0
+for file in "${files[@]}"; do
+  if [[ ! -f "$file" ]]; then
+    printf 'Error: no such file: %s\n' "$file" >&2
+    exit 2
+  fi
+  out="$(scan_file "$file")"
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh]: Fixed in c755ec64. `scan_file`'s exit status is now propagated: a scanner fault (for example a malformed active ERE token that makes awk exit non-zero with no stdout) fails closed with exit 2 instead of the empty output being read as "clean". Added a regression test ("malformed active token exits 2 (fail closed, no silent pass)") that proves the hole is closed. Suite is 14/14.

File: scripts/check-skill-portability.sh [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+  awk '
+    function is_annotated(l) { return l ~ /portability-ok:/ }
+    function is_comment(l) { return l ~ /^[[:space:]]*#/ || l ~ /<!--/ }
+    # Guard markers for the active branch class: a detection-first ladder or a
+    # presence-gated use, which is the correct pattern, not a bare assumption.
+    function is_guarded(l) {
+      return l ~ /origin\/HEAD/ || l ~ /symbolic-ref/ || l ~ /merge-base/ ||
+        l ~ /baseRefName/ || l ~ /[Ff]allback/ || l ~ /[Ff]alling back/ ||
+        l ~ /-> *origin\// || l ~ /if installed/ || l ~ /if using/ ||
+        l ~ /when installed/ || l ~ /when present/
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh]: Fixed in c755ec64. Restricted `is_guarded()` to branch-detection evidence only (`origin/HEAD`, `symbolic-ref`, `merge-base`, `baseRefName`, fallback / falling back, `-> origin/`). Removed the generic optional-dependency presence markers (`if using`, `when installed`, `when present`, `if installed`) that are not branch-resolution evidence — a line like "If using a rebase workflow, diff origin/main first" now correctly fails the gate. Added a regression test proving generic presence prose no longer guards the active branch token. A future class that genuinely needs presence guards adds its own class-scoped markers on enable (the tokens-file staged preamble now states this contract). Suite is 14/14.

File: scripts/check-skill-portability.sh [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+      next
+    }
+    # Pass 2: scan the target file.
+    {
+      line = $0
+      annotated_above = pending_annot
+      if (is_comment(line)) {
+        if (is_annotated(line)) pending_annot = 1
+      } else {
+        pending_annot = 0
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh:179]: Valid finding — deferred, not dropped. Tracked in #611 (cited in this PR's Related section) with the exact fix described here: anchor comment-line recognition to `^[[:space:]]*<!--` so only a standalone comment line can set `pending_annot`, and handle HTML-comment boundaries. Kept out of this PR's ship diff because it adds annotation-carry logic plus test scope beyond the gate's initial ship; the active branch class and the self-test do not exercise it, and the natural annotation forms are unaffected (a same-line `portability-ok:` on the hit line, a standalone comment block directly above, and the whole-file `portability-scope:` escape all still work correctly). Leaving this thread open against #611.

File: scripts/check-skill-portability.sh [hunk 4]
```diff
@@ -0,0 +1,213 @@
...
+    # (a symbolic-ref / merge-base / origin/HEAD ladder that documents an
+    # origin/main last-resort fallback is the correct pattern, not a bare
+    # assumption). Generic optional-dependency presence prose ("if using",
+    # "when present") is NOT branch-resolution evidence and must not sanction a
+    # bare branch default; a future class that genuinely needs presence guards
+    # adds its own class-scoped markers here when it is enabled.
+    function is_guarded(l) {
+      return l ~ /origin\/HEAD/ || l ~ /symbolic-ref/ || l ~ /merge-base/ ||
+        l ~ /baseRefName/ || l ~ /[Ff]allback/ || l ~ /[Ff]alling back/ ||
+        l ~ /-> *origin\//
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh]: Fixed in 8caf57a1. Removed the standalone `[Ff]allback` / `[Ff]alling back` word markers from `is_guarded()`; a co-located branch-resolution command is now the required evidence. Your exact example ("As a fallback for a network timeout, run git diff origin/main") now flags — added as a regression test. Verified against the full corpus: every real detection ladder (review/fanout, review/quality-gate) co-locates a resolution command (`origin/HEAD` / `merge-base` / `-> origin/` / `baseRefName`) on the same line, so all remain guarded and the `--all` audit is unchanged. A legitimately split-across-lines ladder uses the per-site `portability-ok:` escape. Suite is 15/15.

File: scripts/check-skill-portability.sh [hunk 5]
```diff
@@ -0,0 +1,221 @@
...
+  # git pathspec does not match under git's default (non-pathname) globbing.
+  while IFS= read -r f; do
+    case "$f" in
+    plugins/*/skills/*) ;;
+    *) continue ;;
+    esac
+    is_scannable "$f" || continue
+    [[ -f "$f" ]] || continue # a rename-away/deletion leaves nothing to scan
+    files+=("$f")
+  done < <(git diff --name-only --diff-filter=d "$base" -- 'plugins/' | sort -u)
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh]: Fixed in 15d8399be1. The changed-file collection now reads NUL-delimited: `git diff --name-only --diff-filter=d -z`, `sort -z -u`, and a `while IFS= read -r -d ''` loop. Without `-z`, git emitted a non-ASCII path as `"plugins/…/caf\303\251.md"` (quoted, leading `"`), which missed the `plugins/*/skills/*` case glob (and `[[ -f ]]` failed on the quoted string), so the file was silently dropped from the gate — the silent-exclusion class the #532 convention forbids. Paths now arrive verbatim and are gated regardless of the characters in the name.

Added the first diff-mode regression test — the bare-`<base-ref>` file-collection path was previously uncovered. A git-init'd fixture commits one ASCII-named and one non-ASCII-named coupling file, then diffs against the empty base commit. The ASCII hit proves `-z` left the common path intact; two `COUPLING:` lines prove the quoted path was read, not skipped. Confirmed red before the fix (quoted file dropped, one `COUPLING:` line). Suite is 16/16, shellcheck clean.

Deferred, out of this thread's scope: `--all` mode collects via `find … | sort` (no Git quoting involved) and shares the same latent class only for a literal newline in a filename — not portably hostable on the Windows dev path and not part of this changed-file-collection thread. Noting it rather than fixing it here.

File: scripts/check-skill-portability.sh (hunk 3 above)
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh:179]: Control tower ratifies the #611 deferral at merge digest — valid finding, out of the shipped branch-class scope, exact fix recorded on #611.

File: scripts/check-skill-portability.sh [hunk 6]
```diff
@@ -0,0 +1,225 @@
...
+  exit 0
+fi
+
+# scan_file <path> — print `LINE: token -> text` for each unexcused hit.
+scan_file() {
+  local file="$1"
+  # A whole-file declared narrower scope (the inherent-boundary case) excuses
+  # every hit in the file; the declaration is visible in the diff.
+  if grep -qE 'portability-scope:' -- "$file"; then
+    return 0
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh:142]: Valid finding — the `grep -qE 'portability-scope:'` whole-file escape matches the substring anywhere, so prose or a doc example can silently exempt the file. Deferred to #624 (not fixed inline). Rationale: this and the sibling `portability-ok:` nonempty-reason finding both tighten the gate's escape-declaration grammar, and those grammar tokens are the escapes awaiting operator-convention ratification on #531 (items 1-2) — so the rigor fix bundles with that ratification pass rather than another review cycle, keeping this PR's diff scoped to the quoted-path finding it shipped. Control tower ratifies the deferral at merge digest. Resolving here; tracked on #624.

File: scripts/check-skill-portability.sh [hunk 7]
```diff
@@ -0,0 +1,225 @@
...
+  fi
+  awk '
+    function is_annotated(l) { return l ~ /portability-ok:/ }
```
[the remark this answers is not in the document]
Comment (kyle-sexton) [reply scripts/check-skill-portability.sh:145]: Valid finding — `is_annotated` accepts any `portability-ok:` substring including an empty `<!-- portability-ok: -->`, so a bare `origin/main` can pass with no recorded reason. Deferred to #624 (not fixed inline). Rationale: this and the sibling `portability-scope:` standalone-comment finding both tighten the gate's escape-declaration grammar, and those grammar tokens are the escapes awaiting operator-convention ratification on #531 (items 1-2) — so the rigor fix bundles with that ratification pass rather than another review cycle, keeping this PR's diff scoped to the quoted-path finding it shipped. Control tower ratifies the deferral at merge digest. Resolving here; tracked on #624.

**Comment by kyle-sexton** [comment]
Reconciliation flag — concurrent-lane collision on the escape-grammar findings.

A prior lane resolved the two open Codex P2 threads (`portability-scope:` standalone-comment, line 142; `portability-ok:` non-empty reason, line 145) as **deferred to #624**, keeping this PR's diff scoped to the quoted-path finding. Concurrently — before that deferral was visible — I pushed **6c7fe98**, which fixes both inline plus their #611 sibling (`is_comment` anchoring so an inline `<!-- ... -->` tail on a prose line no longer carries an annotation down):

- `portability-scope:` honored only on a standalone comment line with a non-empty reason
- `is_annotated` requires residual non-whitespace reason text (rejects `<!-- portability-ok: -->`)
- 4 regression tests added; suite 20/20, shellcheck clean; `--all` corpus unchanged

These are **active** (not staged-class) code paths in `scan_file`, so the findings are live bypasses in the shipped gate, not future classes.

This leaves a code-vs-disposition inconsistency: the resolved threads read "deferred to #624 / not fixed inline," but head now contains the fix. **Flagging for human / control-tower reconciliation before merge** — either accept the inline fix at 6c7fe98 and rescope #624 (and the #611 tracking item), or revert 6c7fe98 to honor the scoped-diff deferral. I have not touched the resolved threads and am not merging (human-gated).
